src/main/scala/com/fulcrumgenomics/umi/CorrectUmis.scala

```diff
@@ -0,0 +1,229 @@
+                                 ) extends Metric
+}
+
+@clp(group=ClpGroups.Umi, description=
+  """
+    |Corrects UMIs stored in BAM files when a set of fixed UMIs is in use.  If the set of UMIs used in
+    |an experiment is known and is a subset of the possible randomers of the same length, it is possible
+    |to error-correct UMIs prior to grouping reads by UMI.  This tool takes an input BAM with UMIs in a
+    |tag (RX by default) and set of known UMIs (either on the command line or in a file) and produces:
+    |  1. A new BAM with corrected UMIs in the same tag the UMIs were found in
+    |  2. Optionally a set of metrics about the representation of each UMI in the set
+    |  3. Optionally a second BAM file of reads whose UMIs could not be corrected within the specific parameters
+    |
+    |All of the fixed UMIs must be of he same length, and all UMIs in the BAM file must also have the same
+    |length.  Multiple UMIs that are concatenated with hyphens (e.g. AACCAGT-AGGTAGA) are split apart,
+    |corrected individually and then re-assembled.  A read is accepted only if all the UMIs can be corrected.
+    |
+    |Correction is controlled by two parameters:
+    |  1. --max-mismatches controls how many mismatches (no-calls are counted as mismatches) are tolerated
+    |         between a UMI as read and a fixed UMI.
+    |  2. --min-distance controls how many more mismatches the next best hit must have
+    |
+    |For example, with two fixed UMIs AAAAA and CCCCC and max-mismatches=3 and min-distance=2 the
+    |following would happen:
+    |  - AAAAA would match to AAAAA
+    |  - AAGTG would match to AAAAA with three mismatches because CCCCCC has six mismatches and 6 >= 3 + 2
+    |  - AACCA would be rejected because it is 2 mismatches to AAAAA and 3 to CCCCCC and 3 <= 2 + 2
+    |
+    |The set of fixed UMIs may be specified on the command line using --umis umi1 umi2 ... or via one or
+    |more files of UMIs with a single sequence per line using --umi-files umis.txt more_umis.txt
+  """)
+class CorrectUmis
+( @arg(flag="i", doc="Input SAM or BAM file.")  val input: PathToBam,
+  @arg(flag="o", doc="Output SAM or BAM file.") val output: PathToBam,
+  @arg(flag="r", doc="Reject BAM file to save unassigned reads.") val rejects: Option[PathToBam] = None,
+  @arg(flag="M", doc="Metrics file to write.") val metrics: Option[FilePath] = None,
+  @arg(flag="m", doc="Maximum number of mismatches between a UMI and an expected UMI.") val maxMismatches: Int,
+  @arg(flag="d", doc="Minimum distance (in mismatches) to next best UMI.") val minDistance: Int,
+  @arg(flag="u", doc="Expected UMI sequences.", minElements=0) val umis: Seq[String] = Seq.empty,
+  @arg(flag="U", doc="File of UMI sequences, one per line.", minElements=0) val umiFiles: Seq[FilePath] = Seq.empty,
+  @arg(flag="t", doc="Tag in which UMIs are stored.") val umiTag: String = ConsensusTags.UmiBases
+) extends FgBioTool with LazyLogging {
```

Comment:
It absolutely will complain!  It requires that a) all the fixed UMIs be the same length, and then b) all the individual UMIs it examines be the same length as the fixed UMIs.  So you can't have a 6bp UMI and a pair of 3bp UMIs in the same BAM without getting a fairly strong warning/error.